exec/vector/src/main/java/org/apache/drill/exec/record/MaterializedField.java

```diff
@@ -49,39 +54,79 @@ private MaterializedField(String name, MajorType type, LinkedHashSet<Materialize
     this.children = children;
   }
 
+  private MaterializedField(String name, MajorType type, int size) {
+    this(name, type, new LinkedHashSet<>(size));
+  }
+
+  private <T> void copyFrom(Collection<T> source, Function<T, MaterializedField> transformation) {
+    Preconditions.checkState(children.isEmpty());
+    source.forEach(child -> children.add(transformation.apply(child)));
+  }
+
+  public static MaterializedField create(String name, MajorType type) {
+    return new MaterializedField(name, type, 0);
+  }
+
   public static MaterializedField create(SerializedField serField) {
-    LinkedHashSet<MaterializedField> children = new LinkedHashSet<>();
-    for (SerializedField sf : serField.getChildList()) {
-      children.add(MaterializedField.create(sf));
+    MaterializedField field = new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), serField.getChildCount());
+    if (OFFSETS_FIELD.equals(field)) {
+      return OFFSETS_FIELD;
     }
-    return new MaterializedField(serField.getNamePart().getName(), serField.getMajorType(), children);
+    field.copyFrom(serField.getChildList(), MaterializedField::create);
+    return field;
   }
 
-  /**
-   * Create and return a serialized field based on the current state.
-   */
-  public SerializedField getSerializedField() {
-    SerializedField.Builder serializedFieldBuilder = getAsBuilder();
-    for(MaterializedField childMaterializedField : getChildren()) {
-      serializedFieldBuilder.addChild(childMaterializedField.getSerializedField());
+  public MaterializedField copy() {
+    return copy(getName(), getType());
+  }
+
+  public MaterializedField copy(MajorType type) {
+    return copy(name, type);
+  }
+
+  public MaterializedField copy(String name) {
+    return copy(name, getType());
+  }
+
+  public MaterializedField copy(String name, final MajorType type) {
+    if (this == OFFSETS_FIELD) {
```

Comment:
You could play the same trick with the `bits` field. Just need to make sure we use the same name everywhere. Vectors are notoriously inconsistent in things like this.

Reply:
I agree that it will be good to unify `bits` field as well as part of another JIRA/PR.

Reply:
I don't see what is wrong with the semantics here, please elaborate. Do you mean that it is not OK to reuse `OFFSETS_FIELD`, why? Is not it immutable (name, type and children are final)? Or do you mean that other `MaterializedField`s can be re-used as well? In the latter case, extending `OFFSET_FIELD` pattern to other `MaterializedField`s should be done in a separate PR. This PR is limited to fixing existing code, including part of the code that compared `OFFSET_FIELD`.

I don't see a need to rename `copy` to `copySafe` as there is no `copyUnsafe` and one should not exists.

Reply:
Yes, I mean that other instances can be reused. By "wrong semantics", I mean that we have singled out one particular field instance to be "atomic" and reusable. Nothing wrong with that. It is simply awkward:

* Not only is the offsets field reusable, so is the bits field. So, we' see to be making a statement that there is something special about offsets.
* Offsets is reusable because it is fixed size, required. So, if we handle the case for one fixed size, required field, we should handle all of them. Else, again, we're saying that something is special about offsets.
* Further, since the resulting code copies sometimes, but not others, this is not a true "copy" operation. It is a "copy if necessary" operation. Implementation-wise, the "copy if necessary" should either return the same field (if no copy is needed) or call a "copy" method to make an actual copy when needed.

None of this is wrong; the code here works. But, if we are going to the trouble to improve working code, we might want to handle the other cases as well.

Reply:
> Not only is the offsets field reusable, so is the bits field. So, we' see to be making a statement that there is something special about offsets.

Offset field was already handled differently compared to other fields including bits field prior to the PR. The only new behavior that the PR introduces is a singleton pattern for offset field during copy operation to avoid compare by name where compare by identity should be used in the first place. As there was no compare by name for bits, bits field is not handled. I do agree that bits can be handled similar to offset field, but it is not a goal for this PR. The goal of the PR outlined in the JIRA description is to avoid using `clone()` for `MaterializedField`.

> Offsets is reusable because it is fixed size, required. So, if we handle the case for one fixed size, required field, we should handle all of them. Else, again, we're saying that something is special about offsets.

Offset field is reusable because it is immutable (name, type and children are all final). Bits field can be also made reusable, but it is not a goal of the PR. The PR does not make offset field special as it was already handled differently prior to the PR.

>Further, since the resulting code copies sometimes, but not others, this is not a true "copy" operation. It is a "copy if necessary" operation. Implementation-wise, the "copy if necessary" should either return the same field (if no copy is needed) or call a "copy" method to make an actual copy when needed.

It is the only `copy` operation that is available outside of `MaterializedField` class and no any other `copy` should exist. The `copy` decides whether or not it needs to create another instance or it can return a singleton.